src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp

```diff
@@ -440,10 +456,14 @@ void DistributedAsyncInsertDirectoryQueue::processFile(std::string & file_path,
         if (thread_trace_context)
             thread_trace_context->root_span.addAttribute(std::current_exception());
 
+        incrementFilesRetryNum(file_path);
+
         e.addMessage(fmt::format("While sending {}", file_path));
-        if (isDistributedSendBroken(e.code(), e.isRemoteException()))
+        if (isDistributedSendBroken(e.code(), e.isRemoteException()) || (max_retries != 0 && files_retry[file_path] == max_retries))
         {
             markAsBroken(file_path);
+            if (const auto it = files_retry.find(file_path); it != files_retry.end())
+                files_retry.erase(file_path);
```

Comment:
I think we should similarly erase the file path from `files_retry` in `markAsSend`, otherwise `files_retry` might grow without a limit.

Reply:
yes，the file should not be set to broken，it needs to be deleted

Reply:
This is not what I meant. We shouldn't mark files as sent when we haven't sent them. I meant to erase the file from `files_retry` when we mark then sent.

Reply:
I get it, files_retry only records files with exception, and when the number of retries has reached, the file should be deleted instead of being placed in broken

Reply:
OK，I'll change it to broken

Reply:
Now you are removing them. Why? Why not just marking them broken?

Reply:
Because broken data will always exist, It takes up disk space